src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js

```diff
@@ -0,0 +1,198 @@
+ * @constructor
+ */
+e2e.openpgp.pgpmime.MimeNode = function(options, opt_parent) {
+  this.parent = opt_parent || this;
+  this.multipart_ = options.multipart;
+
+  this.children_ = [];
+  this.header_ = /** @type {e2e.openpgp.pgpmime.types.Header} */ ({});
+  this.content_ = null;
+  this.boundary_ = '';
+
+  // set headers
+  if (goog.isDefAndNotNull(options.optionalHeaders)) {
+    goog.array.forEach(options.optionalHeaders, goog.bind(function(header) {
+      if (goog.isDefAndNotNull(header.name) &&
+          goog.isDefAndNotNull(header.value)) {
+        if (goog.isDefAndNotNull(header.params)) {
+          this.setHeader_(header.name, header.value, header.params);
+        } else {
+          this.setHeader_(header.name, header.value);
+        }
+      }
+    }, this));
+  }
+
+  this.setBoundary_();
+};
+
+
+/**
+ * Sets the MIME message boundary for a node.
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setBoundary_ = function() {
+  // TODO: Strictly ensure that the boundary value doesn't coincide with
+  //   any string in the email content and headers.
+  this.boundary_ = '---' + goog.string.getRandomString() +
+      Math.floor(Date.now() / 1000).toString();
+};
+
+
+/**
+ * Adds a child to a MIME node.
+ * @param {{
+ *   multipart: boolean,
+ *   optionalHeaders: (Array<!e2e.openpgp.pgpmime.types.Header>|undefined)
+ * }} options Options to initialize for the node.
+ * @return {e2e.openpgp.pgpmime.MimeNode}
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.addChild = function(options) {
+  var node = new e2e.openpgp.pgpmime.MimeNode(options, this);
+  this.children_.push(node);
+  return node;
+};
+
+
+/**
+ * Sets a MIME header.
+ * @param {string} key Name of the header.
+ * @param {string} value Value of the header, possibly including parameters.
+ * @param {Object.<string, string>=} opt_params Optional dict of additional
+ *   parameters
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setHeader_ =
+    function(key, value, opt_params) {
+  var headerValue = e2e.openpgp.pgpmime.Utils.parseHeaderValue(value);
+  if (opt_params) {
+    headerValue.params = headerValue.params || {};
+    goog.object.extend(headerValue.params, opt_params);
+  }
+  goog.object.set(this.header_, key, headerValue);
+};
+
+
+/**
+ * Sets the content.
+ * @param {(string|!e2e.ByteArray)} content The content to set
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
+  this.content_ = content;
+};
+
+
+/**
+ * Builds an RFC 2822 message from the node and its children.
+ * @return {string}
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.buildMessage = function() {
+  var lines = [];
+  var transferEncoding = this.header_[constants.Mime.CONTENT_TRANSFER_ENCODING];
+  var contentType = this.header_[constants.Mime.CONTENT_TYPE];
+
+  if (!this.header_[constants.Mime.CONTENT_DISPOSITION] &&
+              this.content_ && goog.typeOf(this.content_) === 'string') {
```

Comment:
var contentIsString = goog.typeOf... 

You're checking the type a few times, this way it's easier to read.